Source/Particles/NamedComponentParticleContainer.H

```diff
@@ -0,0 +1,128 @@
+    NamedComponentParticleContainer(
+        amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator> && pc,
+        std::map<std::string, int> p_comps,
+        std::map<std::string, int> p_icomps,
+        std::map<std::string, int> p_rcomps,
+        std::map<std::string, int> p_ricomps)
+    : amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>(std::move(pc)),
+    particle_comps(p_comps),
+    particle_icomps(p_icomps),
+    particle_runtime_comps(p_rcomps),
+    particle_runtime_icomps(p_ricomps) {}
+
+    /* Move operators */
+    NamedComponentParticleContainer ( NamedComponentParticleContainer && ) = default;
+    NamedComponentParticleContainer& operator= ( NamedComponentParticleContainer && ) = default;
+
+    /** Create an empty particle container
+     *
+     * This creates a new NamedComponentParticleContainer with same compile-time
+     * and run-time attributes. But it can change its allocator.
+     *
+     * This function overloads the corresponding function from the parent
+     * class (amrex::ParticleContainer)
+     */
+    template <template<class> class NewAllocator=amrex::DefaultAllocator>
+    NamedComponentParticleContainer<NewAllocator>
+    make_alike () const {
+        auto tmp = NamedComponentParticleContainer<NewAllocator>(
+            amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::template make_alike<NewAllocator>(),
+            particle_comps,
+            particle_icomps,
+            particle_runtime_comps,
+            particle_runtime_icomps);
+
+        return tmp;
+    }
+
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::NumRealComps;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::NumIntComps;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::AddRealComp;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::AddIntComp;
+
+    void AddRealComp (const std::string& name, bool comm=true)
+    {
+        auto search = particle_comps.find(name);
+        if (search == particle_comps.end()) {
+            particle_comps[name] = NumRealComps();
+            particle_runtime_comps[name] = NumRealComps() - PIdx::nattribs;
+            AddRealComp(comm);
+        } else {
+            amrex::Print() << Utils::TextMsg::Info(
+                name + " already exists in particle_comps, not adding.");
+        }
+    }
+
+    void AddIntComp (const std::string& name, bool comm=true)
+    {
+        auto search = particle_icomps.find(name);
+        if (search == particle_icomps.end()) {
+            particle_icomps[name] = NumIntComps();
+            particle_runtime_icomps[name] = NumIntComps() - 0;
+            AddIntComp(comm);
+        } else {
+            amrex::Print() << Utils::TextMsg::Info(
+                name + " already exists in particle_icomps, not adding.");
+        }
+    }
+
+    std::map<std::string, int> getParticleComps () const noexcept { return particle_comps;}
```

Comment:
Do you need sorted keys? Otherwise please use `std::unordered_map`

Reply:
OK, I tried that, but this breaks the code. This is because these functions are **already** used in the rest of the code - since this PR is just moving them from `WarpXParticleContainer` to `NamedParticleContainer` - and the part of the code that calls the function expects a `map`.